supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala

```diff
@@ -47,7 +58,75 @@ case class GenomicsFactory(applicationName: String, authMode: GoogleAuthMode, en
+          case None => IO.raiseError(new RuntimeException(s"Unable to find Google Credential for auth `${vpcConfig.auth.name}`."))
+          case Some(googleCredential) => IO {
+            val auth = googleCredential.createScoped(ResourceManagerAuthScopes)
+
+            val cloudResourceManagerBuilder = new CloudResourceManager
+              .Builder(GoogleAuthMode.httpTransport, GoogleAuthMode.jsonFactory, auth)
+              .build()
+
+            val project = cloudResourceManagerBuilder.projects().get(createPipelineParameters.projectId)
+
+            project.buildHttpRequest()
+          }
+        }
+      }
+
+
+      def projectMetadataResponseToLabels(httpResponse: HttpResponse): IO[ProjectLabels] = {
+        IO.fromEither(decode[ProjectLabels](httpResponse.parseAsString())).handleErrorWith {
+          e => IO.raiseError(new RuntimeException(s"Failed to parse labels from project metadata response from Google Cloud Resource Manager API. " +
+            s"${ExceptionUtils.getMessage(e)}", e))
+        }
+      }
+
+
+      def networkFromLabels(vpcConfig: VirtualPrivateCloudConfiguration, projectLabels: ProjectLabels): Network = {
+        val networkLabelOption = projectLabels.labels.find(l => l._1.equals(vpcConfig.name))
+
+        networkLabelOption match {
+          case Some(networkLabel) =>
+            new Network()
+              .setUsePrivateAddress(createPipelineParameters.runtimeAttributes.noAddress)
+              .setName(VirtualPrivateCloudNetworkPath.format(createPipelineParameters.projectId, networkLabel._2))
+          case None =>
+            jobLogger.debug(s"Project `${createPipelineParameters.projectId}` does not have high security network configured. " +
+              s"Project metadata does not have network label key `${vpcConfig.name}`. Falling back to running the job on default network.")
+            new Network().setUsePrivateAddress(createPipelineParameters.runtimeAttributes.noAddress)
+        }
+      }
+
+
+      def createNetworkWithVPC(vpcConfig: VirtualPrivateCloudConfiguration): IO[Network] = {
+        for {
+          projectMetadataResponse <- projectMetadataRequest(vpcConfig).map(_.executeAsync().get())
+          projectLabels <- projectMetadataResponseToLabels(projectMetadataResponse)
+          networkLabels <- IO(networkFromLabels(vpcConfig, projectLabels))
+        } yield networkLabels
+      }
+
+
+      def createNetwork(): Network = {
+        createPipelineParameters.virtualPrivateCloudConfiguration match {
+          case None => new Network().setUsePrivateAddress(createPipelineParameters.runtimeAttributes.noAddress)
+          case Some(vpcConfig) => {
+            createNetworkWithVPC(vpcConfig).handleErrorWith {
+              e => IO.raiseError(new RuntimeException(s"Failed to create Network object for project `${createPipelineParameters.projectId}`. " +
+                s"Error(s): ${ExceptionUtils.getMessage(e)}", e))
+            }.unsafeRunSync()
```

Comment:
Perhaps a stupid question but.... is it "safe" to do this?

Reply:
I think so. I am assuming it is called `unsafe` because it throws exceptions when anything inside it throws exception, and runs it on blocking thread. This function is ideally supposed to call once in the program, at the very end, but since we don't use IO everywhere, we would need to use it here as well.